### ISSUE TITLE
SC2001 misjudging "simple sed

### ISSUE THREAD

**Issue by automorphism88** [koalaman/shellcheck#1768]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2001
- My shellcheck version (`shellcheck --version` or "online"): 0.7.0 and online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

var1='abc123[]{}().*+?\/&^$'
var2="$(sed 's/[].*+?{}()|$^\[]/\\&/g' <<< "$var1")"
var3="$(sed 's/[].*$^/\[]/\\&/g' <<< "$var1")"
var4="$(sed 's/[&/\]/\\&/g' <<< "$var1")"

echo "var1=$var1"
echo "var2=$var2"
echo "var3=$var3"
echo "var4=$var4"

```

I use all three of the above sed commands in a script and noticed I needed to disable SC2001 only for the one on line 4, 

#### Here's what shellcheck currently says:
```
Line 4:
var2="$(sed 's/[].*+?{}()|$^\[]/\\&/g' <<< "$var1")"
        ^-- SC2001: See if you can use ${variable//search/replace} instead.
```

#### Here's what I wanted or expected to see:

Either nothing, or if SC2001 is triggered by line 4, it should also be triggered by lines 5 and 6. The sed expression in line 4 is the most complicated of the three and the least suitable to replace with bash parameter expansion.

The fact that bash's `${variable//search/replace}` has nothing equivalent to `&` in the sed replacement expression means that to accomplish the same thing as sed is doing (adding a backslash before any of the list of characters in the bracket expression), we would need to do a separate bash parameter expansion for each character we wanted to escape. The more characters we need to escape with a backslash, the worse the problem becomes. Therefore, the current behavior of SC2001 is inconsistent.

**Comment by DenKoren** [comment]
+1 to this issue. 
One more case:

```bash
sed 's/[[:cntrl:]]\[[0-9;]*m//g' <<< "${var}"
```

This construction is designed to remove color modifiers from `${var}` value, but not to touch newlines and so on. Shellcheck thinks this could be implemented in shell parameter expansion, but `\[[0-9;]*m` in shell expansion means not the same as `\[[0-9;]*m` in normal regex (sed).

**Comment by automorphism88** [comment]
> +1 to this issue.
> One more case:
> 
> ```shell
> sed 's/[[:cntrl:]]\[[0-9;]*m//g' <<< "${var}"
> ```
> 
> This construction is designed to remove color modifiers from `${var}` value, but not to touch newlines and so on. Shellcheck thinks this could be implemented in shell parameter expansion, but `\[[0-9;]*m` in shell expansion means not the same as `\[[0-9;]*m` in normal regex (sed).

In this case if you're using bash you should be able to accomplish this without sed, if extglob is enabled, using something like this (using ANSI C quoting `$'\e'` for the escape character, but if you want to use `[[:cntrl:]]` bash also supports character classes):

`var="${var//$'\e['*([0-9;])m/}"`

**Comment by DenKoren** [comment]
Thanks for finding the replacement.I think they are not equivalent: your suggestion will match any symbol between `\e` and last `[0-9]m`. But it seems not so important in this particular case, I can't produce the test case that will drop any part of text between two color modifiers because of this difference. 22.01.2020, 18:08, "automorphism88" <notifications@github.com>:+1 to this issue.One more case:sed 's/[[:cntrl:]]\[[0-9;]*m//g' <<< "${var}"This construction is designed to remove color modifiers from ${var} value, but not to touch newlines and so on. Shellcheck thinks this could be implemented in shell parameter expansion, but \[[0-9;]*m in shell expansion means not the same as \[[0-9;]*m in normal regex (sed).In this case if you're using bash you should be able to accomplish this without sed, if extglob is enabled, using something like this (using ANSI C quoting $'\e' for the escape character, but if you want to use [[:cntrl:]] bash also supports character classes):var="${var//$'\e['*([0-9;])m/}"—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by automorphism88** [comment]
> I think they are not equivalent: your suggestion will match any symbol between `\e` and last `[0-9]m`. But it seems not so important in this particular case, I can't produce the test case that will drop any part of text between two color modifiers because of this difference. 

Not so: `*(pattern)` will match any number of occurrences of the pattern with [extglob](https://www.gnu.org/software/bash/manual/html_node/Pattern-Matching.html) enabled, and nothing else. The `*` has a different meaning when it introduces an extglob than it normally would in bash.

**Comment by DenKoren** [comment]
I didn't know that. Thanks a lot for the info! 25.01.2020, 19:52, "automorphism88" <notifications@github.com>:I think they are not equivalent: your suggestion will match any symbol between \e and last [0-9]m. But it seems not so important in this particular case, I can't produce the test case that will drop any part of text between two color modifiers because of this difference. Not so: *(pattern) will match any number of occurrences of the pattern with extglob enabled, and nothing else. The * has a different meaning when it introduces an extglob than it normally would in bash.—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by koalaman** [comment]
@niclan You can do this with `${ROLES// /$'\n'}`

But yes, this is one of the more dubious suggestions ShellCheck makes, and from the number 2001 you can tell it's one of the very early ones from back when ShellCheck was meant for FAQs/tips&tricks more than day-to-day use. It should be updated to make sure the replacement is simple, if not retired entirely.
